### DISCUSSION TITLE
Allow for external `cache/images` dir for next/image processing

### DISCUSSION THREAD

**Discussion by jflayhart** [vercel/next.js#38066] [Ideas]
### Describe the feature you'd like to request

First, LOVE the `next/image` optimizer. It works great and requires no other setup than simply using `next/image` 👍🏻 

## Problem
The only problem is the limited number of loader options that some of us enterprise companies can't use (ie political red tape, vendor lock-in, what have you). So we're left with the [default loader option of processing/storing optimized images](https://nextjs.org/docs/api-reference/next/image#caching-behavior), which doesn't work well for scaled applications running on 50+ distributed containers. We actually have had to turn image opt OFF with `unoptimized` because there's too much overhead processing of images, since the `cache/images` dir must be repopulated on each pod (eg round robin). Even if we do happen to fill the cache on every pod, after lots of CPU usage, we deploy multiple times a day and those ephemeral containers are wiped restarting the entire process. In the end, we decided we may as well not even have image optimization ☹️ 

## Solution
We can't use one of [external vendors](https://nextjs.org/docs/api-reference/next/image#built-in-loaders) NextJS supports, so we're left with the default loader. In order to support distributed systems, what if we introduced a new `next.config.js` option that allows overriding the default `<distDir>/cache/images` directory to make the cache write/read to an external CDN?

For example,

```js
module.exports = {
  images: {
    loader: 'default',
    cacheDirPath: 'https://cdn.example.com/cache/images',
    cdnAccessKey: 23423432
  },
}
```

This way all default image processing will be written to and read from `https://cdn.example.com/cache/images` to a shared cache dir.

### Describe the solution you'd like

Support distributed web servers by allowing the default image loader to **write to** and **read from** a CDN `https://cdn.example.com/cache/images` to create a shared cache dir. We still want to use nextjs' squooshing capabilities, just send those images to a CDN instead of same server.

For example,

```js
module.exports = {
  images: {
    loader: 'default',
    cacheDirPath: 'https://cdn.example.com/cache/images',
    cdnAccessKey: 23423432  // for uploading
  },
}
```

Now image _could_ be fetched like so:

`https://cdn.example.com/cache/images/some-image&w=1080&q=75.jpg`

and if it also stores as webp:

`https://cdn.example.com/cache/images/some-image&w=1080&q=75.webp`

diff sizes or qualities:

`https://cdn.example.com/cache/images/some-image&w=480&q=75.jpg`

I was thinking I'd leave the params in there as part of filename so it matches the filename in a raw bucket store.

### Describe alternatives you've considered

I've tried implementing a custom loader, doesn't work as I want it to in this case at least I couldn't find anything without being super hacky where I implement my own squoosh/sharp component that processes and uploads images and regenerates the link for me. That just didn't work.

```js
 loader: 'default',
 path: 'https://cdn.example.com/cache/images',
```
AND
```js
 loader: 'custom',
 path: 'https://cdn.example.com/cache/images',
```

**Comment by jflayhart** [comment]
Actually, I took a gander at https://github.com/vercel/next.js/blob/ec4df713525d8af299606161f3b691029603fec9/packages/next/server/image-optimizer.ts and https://github.com/vercel/next.js/blob/57c5e2a0245f6b36663812f88f0de4641ca1e9b9/packages/next/client/image.tsx#L414 and it seems we would have to change a lot with my naive approach above.

What if instead we kept all that the same, but simply took the `/cache/images` dir and uploaded it as is to the provided CDN? Then the config could just be:

```js
 loader: 'default',
 path: 'https://cdn.example.com/cache/images',
 // still need upload access
```

and all images are then fetched on that path but written to `.next/cache/images` which is then uploaded to CDN. I'm just wondering how we can make the change easy and sane enough to support uploading NextJS default squoosh encoding to a CDN.

UPDATE: mm then we could run into race conditions if CDN isn't avail or if access isn't eventually consistent 🤔 Then again, that's on the onus of the developer 🤷🏻 But NextJS could simply fallback to unoptimized image if it doesn't exist in CDN yet.

**Comment by dohomi** [comment]
would it not be possible to run one standalone project which is deployed as an image proxy and has only the core NextJS deployed without any additional logic? Then you can point to that URL like this:
```js
// next.config.js
images:{
  path: 'https://cdn.yourcompany.com/_next/image/'
}
```

**Reply by jflayhart** [reply]
Are you talking about this? https://nextjs.org/docs/api-reference/next.config.js/cdn-support-with-asset-prefix

As far as I know, the default next/image cache behavior only works for _same host_. So not sure how we could use an image proxy, but maybe you can elaborate more on what you're thinking?

**Reply by dohomi** [reply]
@jflayhart no I mean just the `images.path` section of your config, but point it the `[your-deployed-nextjs-project]/_next/image/` because my understanding is that all of the image sources should be fetched from that domain - and the cache I guess should work internally in that deployed project. Sure you have to do some testing - not 100% sure if that works for your use case

**Comment by mattpr** [comment]
So on every deploy the image cache has to be rebuilt?
We prefer to maintain a separate/configurable DATA_DIR that is writable by the app's user rather than allowing the app to write in the deploy directory.  

Is there a reason why the images need to be re-optimized on every deploy or would a persistent (across deploys) DATA_DIR be viable for the cache?

[Currently](https://github.com/vercel/next.js/blob/ec4df713525d8af299606161f3b691029603fec9/packages/next/server/image-optimizer.ts#L194) the cache dir is hard-coded as relative to the distDir.

We also have turned off image optimization because the user that runs our app code does not have write access to the app code directory (built on the CI).

**Comment by ewah** [comment]
i've been premaking my distDir layout and symlink the cache/images directory to an external mount
